docs/specs/devcontainer-reference.md

```diff
@@ -0,0 +1,144 @@
+
+There are two kinds of environment variables:
+Container. These variables are part of the container when it is created and are available at all points in its lifecycle.
+Remote. This variables are to be set by a **development container** supporting tool as part of its environment when it runs. This variables can change during the lifetime of the container.
+
+## Mounts
+
+Mounts allow containers to have access to the underlying machine, share data between containers and to persist information between **development containers**. 
+
+It is important to note that these mounts are from the underlying compute environment and thus in cloud environments might not have access to the same data as a local machine.
+
+## Users
+
+Users control the permissions of applications executed in the containers, allowing the developer to control them. The specification takes into account two types of user definuitions:
+
+Container User. The user that will be used for all operations that run inside a container. It is used to run lifecycle commands among others.
+Remote User. In case the developer wants to use different users for different purposes, the remote user is the user that connecting tools should use to execute operations inside the container.
+
+
+# Lifecycle
+
+A development environment goes through different lifecycle events during its use in the outer and inner loop of development.
+
+- Configuration Validation.
+- Environment Creation.
+- Environment Stop.
+- Environment Resume.
+- Environment Connection.
+
+## Configuration Validation
+
+This section consists of actions required to validate the `devcontainer.json` configuration and does not contain values that conflict with each other.
+
+The tasks executed when validating the configuration are:
+
+- `workspace-folder` is required since it defines where the `devcontainer.json` file is located and identifies the configuration to be used for the environment.
+- Validate access to the `workspace-folder`. 
+- Validate that the metadata (for example `devcontainer.json`) contains all parameters required for the selected configuration type.
+
+## Environment Creation
+
+
+### Initialization
+
+During this step the following is executed:
+- Validate access to the container orchestrator specified by the configuration.
+- Execution of `initializeCommand`.
+
+### Image creation
+
+The first part of an environment creation is generating the final image(s) that the **development containers** are going to use. This step is orchestrator dependent and can consist of just pulling a docker image, running docker build, or docker-compose build. Additionally this step is useful on its own since it permits the creation of intermediate images that can be uploaded and used by other users and thus cut down on creation time. It is encouraged that tools implementing this specification give access to a command that just executes this step.
```

Comment:
```suggestion
The first part of environment creation is generating the final image(s) that the **development containers** are going to use. This step is orchestrator dependent and can consist of just pulling a Docker image, running Docker build, or docker-compose build. Additionally, this step is useful on its own since it permits the creation of intermediate images that can be uploaded and used by other users, thus cutting down on creation time. It is encouraged that tools implementing this specification give access to a command that just executes this step.
```